File: Queue/go/queue.go
package main

import "fmt"

type Queue struct {
	items []int
}

func NewQueue() *Queue {
	return &Queue{items: []int{}}
}

func (q *Queue) isEmpty() bool {
	return len(q.items) == 0
}

func (q *Queue) enqueue(item int) {
	q.items = append(q.items, item)
}

func (q *Queue) dequeue() (int, error) {
	if !q.isEmpty() {
		item := q.items[0]
		q.items = q.items[1:]
		return item, nil
	}
	return 0, fmt.Errorf("Queue is empty")
}

func (q *Queue) peek() (int, error) {
	if !q.isEmpty() {
		return q.items[0], nil
	}
	return 0, fmt.Errorf("Queue is empty")
}

func (q *Queue) size() int {
	return len(q.items)
}

func main() {
	// Create a new queue
	queue := NewQueue()

	// Push elements onto the queue
	queue.enqueue(10)
	queue.enqueue(20)
	queue.enqueue(30)

	// Get the top element (peek)
	if item, err := queue.peek(); err == nil {
		fmt.Println(item) // Output: 30
	}

	// Pop an element from the queue
	if item, err := queue.dequeue(); err == nil {
		fmt.Println(item) // Output: 30
	}

	// Check if the queue is empty
	fmt.Println(queue.isEmpty()) // Output: false

	// Get the current size of the queue
	fmt.Println(queue.size()) // Output: 2
}

File: Queue/java/src/main/java/Queue.java
package src.main.java;
import java.util.ArrayList;

public class Queue<T> {
    private ArrayList<T> items;

    public Queue() {
        items = new ArrayList<>();
    }

    public boolean isEmpty() {
        return items.isEmpty();
    }

    public void enqueue(T item) {
        items.add(item);
    }

    public T dequeue() {
        if (!isEmpty()) {
            return items.remove(0);
        } else {
            throw new IndexOutOfBoundsException("Queue is empty. Cannot dequeue from an empty queue.");
        }
    }

    public T peek() {
        if (!isEmpty()) {
            return items.get(0);
        } else {
            throw new IndexOutOfBoundsException("Queue is empty. Cannot dequeue from an empty queue.");
        }
    }

    public int size() {
        return items.size();
    }

    public static void main(String[] args) {
        // Create a new queue
        Queue<Integer> queue = new Queue<>();

        // Push elements onto the queue
        queue.enqueue(10);
        queue.enqueue(20);
        queue.enqueue(30);

        // Get the first element (peek)
        System.out.println(queue.peek()); // Output: 10

        // Remove an element from the queue
        System.out.println(queue.dequeue()); // Output: 10

        // Check if the queue is empty
        System.out.println(queue.isEmpty()); // Output: false

        // Get the current size of the queue
        System.out.println(queue.size()); // Output: 2
    }
}
